drain/drain.py

```diff
@@ -0,0 +1,418 @@
+            if not first:
+                fee = Millisatoshi(ch['base_fee_millisatoshi'])
+                fee += amount_iter * ch['fee_per_millionth'] // 10**6
+                if fee > amount_iter:
+                    raise RpcError(payload['command'], payload, {'message': 'cannot cover fees to %s %s' % (payload['command'], payload['amount'])})
+                amount_iter -= fee
+            first = False
+            r['msatoshi'] = amount_iter.millisatoshis
+            r['amount_msat'] = amount_iter
+
+
+def get_channel(plugin, payload, peer_id, scid, check_state: bool=False):
+    peer = plugin.rpc.listpeers(peer_id).get('peers')[0]
+    channel = next(c for c in peer['channels'] if 'short_channel_id' in c and c['short_channel_id'] == scid)
+    if check_state:
+        if channel['state'] != "CHANNELD_NORMAL":
+            raise RpcError(payload['command'], payload, {'message': 'Channel %s not in state CHANNELD_NORMAL, but: %s' % (scid, channel['state']) })
+        if not peer['connected']:
+            raise RpcError(payload['command'], payload, {'message': 'Channel %s peer is not connected.' % scid})
+    return channel
+
+
+def spendable_from_scid(plugin, payload, scid):
+    # only fetch funds once to reduce RPC load
+    if not "funds" in payload:
+        payload['funds'] = plugin.rpc.listfunds().get('channels')
+
+    try:
+        channel_funds = next(c for c in payload['funds'] if 'short_channel_id' in c and c['short_channel_id'] == scid)
+    except StopIteration as e:
+        return Millisatoshi(0), Millisatoshi(0)
+    peer_id = channel_funds['peer_id']
+    funds_our = Millisatoshi(channel_funds['our_amount_msat'])
+    try:
+        channel_peer = get_channel(plugin, payload, peer_id, scid, True)
+    except RpcError as e:
+        return Millisatoshi(0), Millisatoshi(0)
+
+    # we check amounts via gossip and not wallet funds, as its more accurate
+    our = Millisatoshi(channel_peer['to_us_msat'])
+    total = Millisatoshi(channel_peer['total_msat'])
+    our_reserve = Millisatoshi(channel_peer['our_reserve_msat'])
+    their_reserve = Millisatoshi(channel_peer['their_reserve_msat'])
+    their = total - our
+
+    # reserves maybe not filled up yet
+    if our < our_reserve:
+        our_reserve = our
+    if their < their_reserve:
+        their_reserve = their
+
+    #spendable = our - our_reserve
+    spendable = channel_peer['spendable_msat']
+    receivable = their - their_reserve
+
+    return spendable, receivable
+
+
+def peer_from_scid(plugin, payload, short_channel_id, my_id):
+    channels = plugin.rpc.listchannels(short_channel_id).get('channels')
+    for ch in channels:
+        if ch['source'] == my_id:
+            return ch['destination']
+    raise RpcError(payload['command'], payload, {'message': 'Cannot find peer for channel: ' + short_channel_id})
+
+
+def find_worst_channel(route):
+    if len(route) < 4:
+        return None
```

Comment:
hmmm, good point. this one was copy&pasta from @gallizoltan rebalance plugin. I think this was done to prevent removing/excluding too many hops if route is already too short or something.

Also, it does not seem to be related to the remaining drain 'Could not find a route' bug.

Reply:
Hm, the `start_id` is 2, meaning it is the third entry in the route...

@gallizoltan can you elaborate on this version of the `find_worst_channel` function maybe (rebalance has the same) ?

Reply:
Okay, it is partly because of historical reasons, and maybe obsolete.

When I wrote this I assumed we need at least 3 channels/hops for a rebalance:
1) The outgoing channel to the outgoing node
2) A middle channel to the incoming node
3) The incoming channel to my node

Number 1 and 3 are fixed. We could only exclude number 2.
But when I wrote this, there was no `setchannelfee` function. A node applied the same fee for all its channels. So there was no point to exclude channel number 2, because its fee set by the outgoing node, so we could not find a cheaper channel instead of that.
This is why `start_id` is 2: possible excludes starts with the third entry in the route.